File: 백준/Silver/9095. 1， 2， 3 더하기/1， 2， 3 더하기.py
N = int(input())

memo = {
    1: 1,
    2: 2,
    3: 4
}

def recursive(n):
    if n in memo:
        return memo[n]
    
    if n not in memo:
        memo[n] = recursive(n-1) + recursive(n-2) + recursive(n-3)

    return memo[n]

for _ in range(N):
    n = int(input())
    print(recursive(n))

File: 백준/Silver/9095. 1， 2， 3 더하기/1， 2， 3 더하기.swift
import Foundation

let n = Int(readLine()!)!

var result = 0

func recursive(target: Int, cur: Int) {
  if cur > target { return }
  if cur == target { result += 1; return }

  for i in 1...3 {
    recursive(target: target, cur: cur + i)
  }
}

for _ in 0..<n {
  let input = Int(readLine()!)!
  recursive(target: input, cur: 0)
  print(result)
  result = 0
}

